### PR TITLE
[Wav2Vec2] Fix normalization for non-padded tensors

### LINKED ISSUE #13504: Wav2vec2Processor normalization issues on transformers 4.10.0
When fine-tuning `facebook/wav2vec2-large-robust-ft-swbd-300h` I noticed I couldn't reproduce past training results from transformers version 4.9.2 now on 4.10. I noticed that inputs are not being correctly normalized with zero mean and unit variance in this new version. This seems to happen when `return_attention_mask=True`, audios in a batch input have different lengths and no padding is done.

## Environment info

- `transformers` version: 4.10.0
- Platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.11
- PyTorch version (GPU?): 1.8.1+cu102 (True)
- Tensorflow version (GPU?): 2.6.0 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help
@patrickvonplaten 
@sgugger

## Information

Model I am using (Bert, XLNet ...): Wav2Vec 2.0

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

1. Load Wav2Vec2Processor from `facebook/wav2vec2-large-robust-ft-swbd-300h`
2. Call processor with batched inputs of individual different lengths

Sample code to replicate the error:
```
import numpy as np
from transformers import Wav2Vec2Processor

processor = Wav2Vec2Processor.from_pretrained("facebook/wav2vec2-large-robust-ft-swbd-300h")

sample_rate = 16000
length_1 = 10
length_2 = 20

# Generate dummy input audios of same sample rate but different lengths
input_1 = np.random.rand((sample_rate * length_1))
input_2 = np.random.rand((sample_rate * length_1))
input_3 = np.random.rand((sample_rate * length_2))
 
same_length_result = processor([input_1, input_2], sampling_rate=sample_rate)
different_length_result = processor([input_1, input_3], sampling_rate=sample_rate)

# Show normalized batched audios when using same length
print(same_length_result)
# Show normalized batched audios when using different length
print(different_length_result)

# Check same audio suffers different transformations according to length of audios in batch
np.testing.assert_array_equal(same_length_result["input_values"][0], different_length_result["input_values"][0])
```

## Expected behavior
A successful assert. Both processed inputs should be equal, with a mean close to 0 and a standard deviation close to 1.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR fixes a problem with normalization when the input is a list of different length that is not numpified - see: https://github.com/huggingface/transformers/issues/13504

Just noticed that this bug is pretty severe actually as it affects all large-Wav2Vec2 fine-tuning  :-/.
It was introduced by me in this PR: https://github.com/huggingface/transformers/pull/12804/files - I should have written more and better tests for this. 

=> This means that from transformers 4.9.0 to until this PR is merged the normalization for all large Wav2Vec2 models was way off when fine-tuning the model.
 
@LysandreJik - do you think it might be possible to do a patched release for this?

### CODE REVIEW

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 1]
```diff
@@ -79,13 +79,27 @@ def __init__(
...
+        input_values: List[np.ndarray], attention_mask: List[np.ndarray], padding_value=0.0
+    ) -> List[np.ndarray]:
         """
         Every array in the list is normalized to have zero mean and unit variance
         """
-        normed_input_values = [
-            (x - np.mean(x[:i])) / np.sqrt(np.var(x[:i]) + 1e-5) for x, i in zip(input_values, input_lengths)
-        ]
+
+        if attention_mask is not None and any(0 in mask for mask in attention_mask):
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:89]:
```suggestion
        if attention_mask is not None:
```

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 2]
```diff
@@ -79,13 +79,27 @@ def __init__(
...
+                [
+                    (x - np.mean(x[:i])) / np.sqrt(np.var(x[:i]) + 1e-7)
+                    for x, i in zip(input_values, attention_mask.sum(-1))
+                ]
+            )
+            normed_input_values[(1 - attention_mask).astype(np.bool)] = padding_value
+        else:
+            normed_input_values = np.asarray([(x - np.mean(x)) / np.sqrt(np.var(x) + 1e-7) for x in input_values])
+
+        if isinstance(input_values[0], np.ndarray):
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:100]:
```suggestion
```

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 3]
```diff
@@ -79,13 +79,27 @@ def __init__(
...
+            input_values = [x.astype(np.float32) for x in input_values]
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:101]:
```suggestion
```

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 4]
```diff
@@ -79,13 +79,20 @@ def __init__(
         self.do_normalize = do_normalize
 
     @staticmethod
-    def zero_mean_unit_var_norm(input_values: List[np.ndarray], input_lengths: List[int]) -> List[np.ndarray]:
+    def zero_mean_unit_var_norm(
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:82]: The responsibility of retrieving the correct length from the attention mask should be in this method since `input_values` and `attention_mask` are the well-known inputs to functions in `transformers`

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 5]
```diff
@@ -196,19 +195,33 @@ def __call__(
             return_attention_mask=return_attention_mask,
         )
 
-        if "attention_mask" in padded_inputs:
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:199]: This part is removed/cleaned-up

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 6]
```diff
@@ -172,14 +179,6 @@ def __call__(
             and (isinstance(raw_speech[0], np.ndarray) or isinstance(raw_speech[0], (tuple, list)))
         )
 
-        # make sure input is in list format
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:175]: Currently all the padding is happening in pure python and not in numpy so let's move the numpification further down

File: tests/test_feature_extraction_wav2vec2.py [hunk 7]
```diff
@@ -134,7 +134,22 @@ def _check_zero_mean_unit_variance(input_vector):
         _check_zero_mean_unit_variance(input_values[1, :1000])
         _check_zero_mean_unit_variance(input_values[2])
 
-    def test_zero_mean_unit_variance_normalization_trunc(self):
+    def test_zero_mean_unit_variance_normalization(self):
```
Comment (patrickvonplaten) [inline tests/test_feature_extraction_wav2vec2.py:137]: Add test to make sure normalization always works as expected

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 8]
```diff
@@ -79,13 +79,20 @@ def __init__(
         self.do_normalize = do_normalize
 
     @staticmethod
-    def zero_mean_unit_var_norm(input_values: List[np.ndarray], input_lengths: List[int]) -> List[np.ndarray]:
+    def zero_mean_unit_var_norm(
+        input_values: List[np.ndarray], attention_mask: List[np.ndarray], padding_value=0.0
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:83]: seems `padding_value` is not used in this method

File: tests/test_feature_extraction_wav2vec2.py [hunk 9]
```diff
@@ -134,7 +134,22 @@ def _check_zero_mean_unit_variance(input_vector):
...
+        feat_extract = self.feature_extraction_class(**self.feat_extract_tester.prepare_feat_extract_dict())
+        lengths = range(800, 1400, 200)
+        speech_inputs = [floats_list((1, x))[0] for x in lengths]
+        processed = feat_extract(speech_inputs)
+        input_values = processed.input_values
+
+        def _check_zero_mean_unit_variance(input_vector):
+            self.assertTrue(np.abs(np.mean(input_vector)) < 1e-3)
+            self.assertTrue(np.abs(np.var(input_vector) - 1) < 1e-3)
```
Comment (patil-suraj) [inline tests/test_feature_extraction_wav2vec2.py:146]: Great test!

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 10]
```diff
@@ -79,13 +79,20 @@ def __init__(
...
-        normed_input_values = [
-            (x - np.mean(x[:i])) / np.sqrt(np.var(x[:i]) + 1e-5) for x, i in zip(input_values, input_lengths)
-        ]
+
+        if attention_mask is not None:
+            normed_input_values = [
+                (x - x[:i].mean()) / np.sqrt(x[:i].var() + 1e-7) for x, i in zip(input_values, attention_mask.sum(-1))
+            ]
+        else:
+            normed_input_values = [(x - x.mean()) / np.sqrt(x.var() + 1e-7) for x in input_values]
```
Comment (patil-suraj) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:94]: Nice!
This will also affect `Speech2TextFeatureExtractor`, no ? It also does normalization 

https://github.com/huggingface/transformers/blob/09549aa18c7c1b56106790a04c06566846de209b/src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py#L95-L108

**Review by patil-suraj — APPROVED** [review]
Great catch!

This looks good to me.

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py (hunk 10 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:94]: Actually yeah that's a good point - will check!

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 11]
```diff
@@ -196,19 +193,33 @@ def __call__(
...
         # zero-mean and unit-variance normalization
         if self.do_normalize:
+            attention_mask = (
+                np.array(attention_mask, dtype=np.bool)
+                if self._get_padding_strategies(padding, max_length=max_length) is not PaddingStrategy.DO_NOT_PAD
+                else None
+            )
             padded_inputs["input_values"] = self.zero_mean_unit_var_norm(
-                padded_inputs["input_values"], input_lengths=input_lengths
+                padded_inputs["input_values"], attention_mask=attention_mask, padding_value=self.padding_value
```
Comment (LysandreJik) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:222]: This method doesn't have a `padding_value` attribute

**Review by LysandreJik — APPROVED** [review]
Seems to look good but will delegate to @patil-suraj and @anton-l's w2v2 knowledge.

Let me know once this is merged so that I may release a patch.

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 12]
```diff
@@ -196,19 +193,33 @@ def __call__(
...
+        # convert attention_mask to correct format
+        attention_mask = padded_inputs.get("attention_mask")
+        if attention_mask is not None:
+            padded_inputs["attention_mask"] = [np.asarray(array, dtype=np.bool) for array in attention_mask]
 
         # zero-mean and unit-variance normalization
         if self.do_normalize:
+            attention_mask = (
+                np.array(attention_mask, dtype=np.bool)
+                if self._get_padding_strategies(padding, max_length=max_length) is not PaddingStrategy.DO_NOT_PAD
```
Comment (anton-l) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:218]: When `padding==True` and `attention_mask==None`, this makes `attention_mask=np.array([False])` and breaks iteration in `zero_mean_unit_var_norm` later on.

**Review by anton-l — CHANGES_REQUESTED** [review]
LGTM other than the small issues already pointed out, thanks for fixing it!

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py (hunk 12 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:218]: Great catch! It shouldn't really happen that padding is enabled and no attention mask is returned, but definitely better to be save here!

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py (hunk 10 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:92]: Ok there were a couple of problems here actually :-/ Think most of them are fixed now

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py (hunk 8 above)
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py]: It should be used ;-)

File: src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py [hunk 13]
```diff
@@ -107,15 +110,21 @@ def utterance_cmvn(
             std = np.sqrt(np.maximum(var, 1e-10))
             x = np.divide(x, std)
 
+        if x.shape[0] > input_length:
+            x[input_length:] = padding_value
```
Comment (patrickvonplaten) [inline src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:114]: @patil-suraj - In the beginning conv layers are used, so IMO it's important to set incorrectly normalized values accordingly

File: tests/test_feature_extraction_wav2vec2.py [hunk 14]
```diff
@@ -120,21 +120,45 @@ def test_call(self):
...
+    def test_zero_mean_unit_variance_normalization_np(self):
         feat_extract = self.feature_extraction_class(**self.feat_extract_tester.prepare_feat_extract_dict())
         speech_inputs = [floats_list((1, x))[0] for x in range(800, 1400, 200)]
-        processed = feat_extract(speech_inputs, padding="longest", return_tensors="np")
-        input_values = processed.input_values
 
-        def _check_zero_mean_unit_variance(input_vector):
-            self.assertTrue(np.abs(np.mean(input_vector)) < 1e-3)
-            self.assertTrue(np.abs(np.var(input_vector) - 1) < 1e-3)
+        paddings = ["longest", "max_length", "do_not_pad"]
```
Comment (patrickvonplaten) [inline tests/test_feature_extraction_wav2vec2.py:127]: normalization is checked now for all kinds of paddings

File: tests/test_feature_extraction_speech_to_text.py [hunk 15]
```diff
@@ -136,18 +136,49 @@ def test_call(self):
...
+            _check_zero_mean_unit_variance(input_features[1][: fbank_feat_lengths[1]], var_tol)
+            _check_zero_mean_unit_variance(input_features[2][: fbank_feat_lengths[2]], var_tol)
+
+    def test_cepstral_mean_and_variance_normalization_np(self):
+        feature_extractor = self.feature_extraction_class(**self.feat_extract_tester.prepare_feat_extract_dict())
+        speech_inputs = [floats_list((1, x))[0] for x in range(800, 1400, 200)]
+
+        paddings = ["longest", "max_length", "do_not_pad"]
+        max_lengths = [None, 16, None]
+        var_tolerances = [1e-3, 1e-3, 1e-1]
```
Comment (patrickvonplaten) [inline tests/test_feature_extraction_speech_to_text.py:166]: raw python float input differs significantly from numpy so that the variance tolerance has to be much lower here

File: src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py (hunk 13 above)
Comment (patil-suraj) [reply src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:114]: Good point, yeah! Just checked in fairseq, padding happens after the processing. So this is important here.

**Review by anton-l — APPROVED** [review]
All slow tests now pass for Wav2Vec and Hubert, nice!

**Review by patil-suraj — APPROVED** [review]
LGTM! Thanks for adding all those tests :)
